<a href="https://colab.research.google.com/github/peremartra/llama-glu-expansion-pruning/blob/main/notebooks/00_Neuron_Selection_Method_Comparison.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# GLU Pruning Research
## 00 - Neuron Selection Method Comparison: PPM vs L2 vs VOW vs PON

### Exploring GLU Expansion Ratios in Llama-3.2 Models
by [Pere Martra](https://github.com/peremartra)

[![Paper](https://img.shields.io/badge/arxiv-Paper-blue?logo=osf&logoColor=white)](https://www.arxiv.org/abs/2512.22671)
[![GitHub](https://img.shields.io/badge/⭐_Star-OptiPFair-orange?logo=github&logoColor=white)](https://github.com/peremartra/optipfair)
[![PyPI](https://img.shields.io/pypi/v/optipfair?logo=python&logoColor=white&label=v)](https://pypi.org/project/optipfair/)

**Repository:** [github.com/peremartra/llama-glu-expansion-pruning](https://github.com/peremartra/llama-glu-expansion-pruning)

---

**Colab Environment:** GPU L4

**Models:**
* Llama-3.2-1B (base)

---

## 📋 Notebook Objective

This preliminary experiment compares three neuron importance metrics for GLU-based pruning:
- **PPM** (Peak-toPeak Magnitude) - Default method in OptiPFair

```
gate_max_abs = torch.max(gate_weight, dim=1).values + torch.abs(torch.min(gate_weight, dim=1).values)
up_max_abs = torch.max(up_weight, dim=1).values + torch.abs(torch.min(up_weight, dim=1).values)
importance_scores = gate_max_abs + up_max_abs
```

- **L2** (L2 Norm)

```
gate_norms = torch.norm(gate_weight, p=2, dim=1)
up_norms = torch.norm(up_weight, p=2, dim=1)
importance_scores = gate_norms + up_norms
```

- **VOW** (Variance of Weights)
```
gate_variance = torch.var(gate_weight, dim=1)
up_variance = torch.var(up_weight, dim=1)
importance_scores = gate_variance + up_variance
```
- **PON** (Product of Norms)
```
gate_norms = torch.norm(gate_weight, p=1, dim=1)
up_norms = torch.norm(up_weight, p=1, dim=1)
importance_scores = gate_norms * up_norms
```

**Goal:** Empirically determine which method best preserves model capabilities under pruning before conducting the main experiments. We will create three versions of Llama-3.2-1B pruned at 10% (one per method) and evaluate them sequentially on progressively demanding benchmarks.

**Evaluation Strategy:** We'll start with the fastest benchmarks (BoolQ, Lambada) to quickly identify clear winners or losers. If methods show similar performance, we'll proceed to more demanding benchmarks (IFEval, GSM8K) for differentiation.

**Why this matters:** The choice of neuron selection method fundamentally affects which neurons are removed during pruning. This "tournament" ensures our main experiments use the most effective approach, with results documented in the paper's methodology section.

**Output:** The winning method will be used exclusively for all subsequent pruning experiments (01 and 02 notebooks).

---

## 📦 Required Libraries

```python
# Install OptiPFair for GLU pruning
!pip install optipfair

# Install LM Evaluation Harness for benchmarking
!pip install lm-eval
```

---

**Note:** This notebook is part of a larger research project on width pruning in GLU architectures. See the full paper and codebase at the links above.

---

# Setup & Install

In [1]:
# Install required libraries.
!pip install -q optipfair
!pip install -q lm-eval
!pip install -q langdetect

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.8/62.8 kB 5.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.9/58.9 kB 5.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.9/8.9 MB 143.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 9.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100.8/100.8 kB 9.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 91.1/91.1 kB 7.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 981.5/981.5 kB 59.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done


In [2]:
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from optipfair import prune_model
from datasets import load_dataset
import copy
import json
import gc
from datetime import datetime
from lm_eval import evaluator
from lm_eval.models.huggingface import HFLM

In [3]:
def clear_gpu_cache():
    """Limpia completamente la cache de GPU"""
    if torch.cuda.is_available():
        torch.cuda.empty_cache()
        torch.cuda.synchronize()
    gc.collect()


In [4]:
MODEL_ID = "meta-llama/Llama-3.2-1B"
PRUNING_PERCENTAGE = 10
PERPLEXITY_SAMPLES=500
DEVICE = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

In [6]:
def set_seed(seed=42):
    """Set random seed for reproducibility"""
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)

set_seed(42)
print("✓ Random seed set to 42")

✓ Random seed set to 42


# Load Base Model

In [7]:
base_model = AutoModelForCausalLM.from_pretrained(
  MODEL_ID,
  dtype=torch.float16,
  device_map="auto"
)

config.json:   0%|          | 0.00/843 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.47G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/146 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/185 [00:00<?, ?B/s]

In [8]:
tokenizer = AutoTokenizer.from_pretrained(MODEL_ID)
if tokenizer.pad_token is None:
  tokenizer.pad_token = tokenizer.eos_token

tokenizer_config.json:   0%|          | 0.00/50.5k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/9.09M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/301 [00:00<?, ?B/s]

# Evaluations

## Evaluations Code

In [9]:
def model_evaluation(model_obj, tokenizer, tasks, limit=None):
    """
    Runs lm-eval on a model and tokenizer object already in memory.

    Args:
        model_obj: The PyTorch model object to evaluate.
        tokenizer_obj: The tokenizer object.
        tasks (list): A list of task names for lm-eval.
        limit (int): The number of samples per task.
    """
    model_name = getattr(model_obj.config, '_name_or_path', 'unknown')
    limit_str = f"(limit={limit})" if limit else "(full dataset)"
    print(f"\n{'='*70}")
    print(f"Starting lm-eval on model '{model_name}'")
    print(f"Tasks: {tasks} {limit_str}")
    print(f"{'='*70}\n")

    # Wrap the local model object and tokenizer for lm-eval
    model_wrapper = HFLM(
        pretrained=model_obj,
        tokenizer=tokenizer,
        device=str(DEVICE)
    )

    # Run evaluation
    results = evaluator.simple_evaluate(
        model=model_wrapper,
        tasks=tasks,
        num_fewshot=0,
        limit=limit,
        device=str(DEVICE)
    )

    # Format results for clean display
    formatted_results = {}
    for task_name, res in results["results"].items():
        # Extract relevant metrics based on task type
        if 'perplexity,none' in res:
            # For perplexity tasks (wikitext)
            formatted_results[task_name] = {
                'perplexity': f"{res.get('perplexity,none', 0):.2f}",
                'word_perplexity': f"{res.get('word_perplexity,none', 0):.2f}",
                'bits_per_byte': f"{res.get('bits_per_byte,none', 0):.4f}"
            }
        elif 'acc,none' in res:
            # For accuracy tasks (boolq, etc.)
            formatted_results[task_name] = {
                'accuracy': f"{res.get('acc,none', 0):.4f}",
                'acc_norm': f"{res.get('acc_norm,none', 0):.4f}"
            }
        else:
            # Fallback: store all available metrics
            formatted_results[task_name] = {k: f"{v:.4f}" for k, v in res.items() if isinstance(v, (int, float))}

    return formatted_results

In [10]:
TASKS = ["wikitext", "boolq", "lambada_openai", "piqa"]


In [11]:
BASE_results = model_evaluation(
  model_obj=base_model,
  tokenizer=tokenizer,
  tasks=TASKS,
  limit=None  # Full dataset
)


Starting lm-eval on model 'meta-llama/Llama-3.2-1B'
Tasks: ['wikitext', 'boolq', 'lambada_openai', 'piqa'] (full dataset)



README.md:   0%|          | 0.00/8.76k [00:00<?, ?B/s]

wikitext-2-raw-v1/wikitext-2-raw-v1-trai(…):   0%|          | 0.00/6.18M [00:00<?, ?B/s]

wikitext-2-raw-v1/wikitext-2-raw-v1-vali(…):   0%|          | 0.00/641k [00:00<?, ?B/s]

wikitext-2-raw-v1/wikitext-2-raw-v1-test(…):   0%|          | 0.00/715k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/629 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/60 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/62 [00:00<?, ? examples/s]

README.md:   0%|          | 0.00/19.7k [00:00<?, ?B/s]

boolq/train-00000-of-00001.parquet:   0%|          | 0.00/3.85M [00:00<?, ?B/s]

boolq/validation-00000-of-00001.parquet:   0%|          | 0.00/1.31M [00:00<?, ?B/s]

boolq/test-00000-of-00001.parquet:   0%|          | 0.00/1.31M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/9427 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/3270 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/3245 [00:00<?, ? examples/s]

README.md:   0%|          | 0.00/5.47k [00:00<?, ?B/s]

default/test/default.parquet:   0%|          | 0.00/1.16M [00:00<?, ?B/s]

Generating test split:   0%|          | 0/5153 [00:00<?, ? examples/s]

piqa_train.parquet:   0%|          | 0.00/2.64M [00:00<?, ?B/s]

piqa_validation.parquet:   0%|          | 0.00/300k [00:00<?, ?B/s]

piqa_test.parquet:   0%|          | 0.00/496k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/16113 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/1838 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/3084 [00:00<?, ? examples/s]

Running loglikelihood requests: 100%|██████████| 15369/15369 [04:10<00:00, 61.24it/s]


bootstrapping for stddev: perplexity


100%|██████████| 100/100 [00:08<00:00, 12.12it/s]


In [12]:
BASE_results

{'wikitext': {'sample_len': '62.0000',
  'word_perplexity,none': '11.9568',
  'byte_perplexity,none': '1.5904',
  'bits_per_byte,none': '0.6694'},
 'boolq': {'accuracy': '0.6382', 'acc_norm': '0.0000'},
 'lambada_openai': {'perplexity': '5.42',
  'word_perplexity': '0.00',
  'bits_per_byte': '0.0000'},
 'piqa': {'accuracy': '0.7519', 'acc_norm': '0.7476'}}

## PPM selection model
Peak-to-Peak Magnitude.



In [13]:
# Create Pruned Model with PPM
pruned_model_ppm, stats = prune_model(
    model=copy.deepcopy(base_model),
    pruning_type="MLP_GLU",
    neuron_selection_method="PPM",
    pruning_percentage=PRUNING_PERCENTAGE,
    show_progress=True,
    return_stats=True
)
stats

Pruning all layers: 100%|██████████| 16/16 [00:06<00:00,  2.66it/s]


{'original_parameters': 1235814400,
 'pruned_parameters': 1155303424,
 'reduction': 80510976,
 'percentage_reduction': 6.5148112855781575,
 'expansion_rate': 360.009765625}

In [14]:
PPM_results = model_evaluation(
  model_obj=pruned_model_ppm,
  tokenizer=tokenizer,
  tasks=TASKS,
  limit=None  # Full dataset
)


Starting lm-eval on model 'meta-llama/Llama-3.2-1B'
Tasks: ['wikitext', 'boolq', 'lambada_openai', 'piqa'] (full dataset)



Running loglikelihood requests: 100%|██████████| 15369/15369 [04:18<00:00, 59.43it/s]


bootstrapping for stddev: perplexity


100%|██████████| 100/100 [00:08<00:00, 11.87it/s]


In [15]:
PPM_results

{'wikitext': {'sample_len': '62.0000',
  'word_perplexity,none': '18.3716',
  'byte_perplexity,none': '1.7235',
  'bits_per_byte,none': '0.7853'},
 'boolq': {'accuracy': '0.6220', 'acc_norm': '0.0000'},
 'lambada_openai': {'perplexity': '17.89',
  'word_perplexity': '0.00',
  'bits_per_byte': '0.0000'},
 'piqa': {'accuracy': '0.7242', 'acc_norm': '0.7187'}}

In [16]:
#Delete PPM model
del pruned_model_ppm
clear_gpu_cache()

# L2 Selection Model




In [17]:
# Create Pruned Model with L2
pruned_model_L2, stats = prune_model(
    model=copy.deepcopy(base_model),
    pruning_type="MLP_GLU",
    neuron_selection_method="L2",
    pruning_percentage=PRUNING_PERCENTAGE,
    show_progress=True,
    return_stats=True
)
stats

Pruning all layers: 100%|██████████| 16/16 [00:05<00:00,  2.75it/s]


{'original_parameters': 1235814400,
 'pruned_parameters': 1155303424,
 'reduction': 80510976,
 'percentage_reduction': 6.5148112855781575,
 'expansion_rate': 360.009765625}

In [18]:
L2_results = model_evaluation(
  model_obj=pruned_model_L2,
  tokenizer=tokenizer,
  tasks=TASKS,
  limit=None  # Full dataset
)


Starting lm-eval on model 'meta-llama/Llama-3.2-1B'
Tasks: ['wikitext', 'boolq', 'lambada_openai', 'piqa'] (full dataset)



Running loglikelihood requests: 100%|██████████| 15369/15369 [04:19<00:00, 59.31it/s]


bootstrapping for stddev: perplexity


100%|██████████| 100/100 [00:08<00:00, 12.32it/s]


In [19]:
L2_results

{'wikitext': {'sample_len': '62.0000',
  'word_perplexity,none': 'inf',
  'byte_perplexity,none': 'inf',
  'bits_per_byte,none': 'inf'},
 'boolq': {'accuracy': '0.6232', 'acc_norm': '0.0000'},
 'lambada_openai': {'perplexity': '551.10',
  'word_perplexity': '0.00',
  'bits_per_byte': '0.0000'},
 'piqa': {'accuracy': '0.6518', 'acc_norm': '0.6605'}}

In [20]:
#Delete L2 model
del pruned_model_L2
clear_gpu_cache()

## PON Selection Model


In [21]:
# Create Pruned Model with PON
pruned_model_pon, stats = prune_model(
    model=copy.deepcopy(base_model),
    pruning_type="MLP_GLU",
    neuron_selection_method="PON",
    pruning_percentage=PRUNING_PERCENTAGE,
    show_progress=True,
    return_stats=True
)
stats

Pruning all layers: 100%|██████████| 16/16 [00:05<00:00,  2.70it/s]


{'original_parameters': 1235814400,
 'pruned_parameters': 1155303424,
 'reduction': 80510976,
 'percentage_reduction': 6.5148112855781575,
 'expansion_rate': 360.009765625}

In [22]:
PON_results = model_evaluation(
  model_obj=pruned_model_pon,
  tokenizer=tokenizer,
  tasks=TASKS,
  limit=None  # Full dataset
)


Starting lm-eval on model 'meta-llama/Llama-3.2-1B'
Tasks: ['wikitext', 'boolq', 'lambada_openai', 'piqa'] (full dataset)



Running loglikelihood requests: 100%|██████████| 15369/15369 [04:18<00:00, 59.53it/s]


bootstrapping for stddev: perplexity


100%|██████████| 100/100 [00:08<00:00, 11.94it/s]


In [23]:
PON_results

{'wikitext': {'sample_len': '62.0000',
  'word_perplexity,none': 'inf',
  'byte_perplexity,none': 'inf',
  'bits_per_byte,none': 'inf'},
 'boolq': {'accuracy': '0.6211', 'acc_norm': '0.0000'},
 'lambada_openai': {'perplexity': '1522.95',
  'word_perplexity': '0.00',
  'bits_per_byte': '0.0000'},
 'piqa': {'accuracy': '0.6551', 'acc_norm': '0.6600'}}

In [24]:
del pruned_model_pon
clear_gpu_cache()

## VOW Selection Model


In [25]:
# Create Pruned Model with PON
pruned_model_vow, stats = prune_model(
    model=copy.deepcopy(base_model),
    pruning_type="MLP_GLU",
    neuron_selection_method="VOW",
    pruning_percentage=PRUNING_PERCENTAGE,
    show_progress=True,
    return_stats=True
)
stats

Pruning all layers: 100%|██████████| 16/16 [00:05<00:00,  2.77it/s]


{'original_parameters': 1235814400,
 'pruned_parameters': 1155303424,
 'reduction': 80510976,
 'percentage_reduction': 6.5148112855781575,
 'expansion_rate': 360.009765625}

In [26]:
VOW_results = model_evaluation(
  model_obj=pruned_model_vow,
  tokenizer=tokenizer,
  tasks=TASKS,
  limit=None  # Full dataset
)


Starting lm-eval on model 'meta-llama/Llama-3.2-1B'
Tasks: ['wikitext', 'boolq', 'lambada_openai', 'piqa'] (full dataset)



Running loglikelihood requests: 100%|██████████| 15369/15369 [04:17<00:00, 59.67it/s]


bootstrapping for stddev: perplexity


100%|██████████| 100/100 [00:08<00:00, 12.05it/s]


In [27]:
VOW_results

{'wikitext': {'sample_len': '62.0000',
  'word_perplexity,none': '58.0275',
  'byte_perplexity,none': '2.1370',
  'bits_per_byte,none': '1.0956'},
 'boolq': {'accuracy': '0.6229', 'acc_norm': '0.0000'},
 'lambada_openai': {'perplexity': '407.21',
  'word_perplexity': '0.00',
  'bits_per_byte': '0.0000'},
 'piqa': {'accuracy': '0.6534', 'acc_norm': '0.6654'}}

In [28]:
del pruned_model_vow
clear_gpu_cache()

# 📊 Method Comparison Results

| Model | WikiText PPL ↓ | Lambada PPL ↓ | BoolQ Acc ↑ | PIQA Acc ↑ | Status |
| :--- | :--- | :--- | :--- | :--- | :--- |
| Base | 11.96 | 5.42 | 0.6382 | 0.7519 | Baseline |
| PPM | 18.37 (+54%) | 17.89 (+230%) | 0.6220 (-3%) | 0.7242 (-4%) | ✅ SELECTED |
| VOW | 58.03 (+385%) | 407.21 (+7,413%) | 0.6229 (-2%) | 0.6534 (-13%) | ❌ Rejected |
| L2 | inf | 551.10 (+10,068%) | 0.6232 (-2%) | 0.6518 (-13%) | ❌ Rejected |
| PON | inf | 1522.95 (+28,000%) | 0.6211 (-3%) | 0.6551 (-13%) | ❌ Rejected |


**Conclusion**: PPM is the clear winner with acceptable degradation (~50% PPL increase) while VOW/PON or even the standard L2 show catastrophic failures (10,000%+ increases). PPM will be used for all experiments.

The results are conclusive. While all pruning methods introduce some performance degradation compared to the base model, the PPM (Peak-to-Peak Magnitude) method preserves the model's capabilities far more effectively than the alternatives.

PPM shows only a minor increase in perplexity and a negligible drop in accuracy, demonstrating its ability to remove parameters without causing significant damage to the model's core fluency and comprehension.

VOW, PON and L2 result in a catastrophic loss of performance, especially visible in the wikitext and lambada_openai perplexity scores. A Lambada perplexity score above 2000 for the PON method indicates that the model has become almost unusable for text generation tasks.

This data provides a strong empirical justification for selecting PPM as the sole neuron selection method for all subsequent, large-scale pruning experiments in this research project.

These experiments were powered by **OptiPFair**. If this research helps your work, consider:
- ⭐ Starring [the repo](https://github.com/peremartra/optipfair)
- 📖 Reading the [documentation](https://peremartra.github.io/optipfair/)
- 🐛 Reporting issues or suggesting features